controllers/topology/desired_state.go

```diff
@@ -173,14 +173,49 @@ func computeControlPlane(_ context.Context, s *scope.Scope, infrastructureMachin
 	}
 
 	// Sets the desired Kubernetes version for the control plane.
-	// TODO: improve this logic by adding support for version upgrade component by component
-	if err := contract.ControlPlane().Version().Set(controlPlane, s.Blueprint.Topology.Version); err != nil {
+	version, err := computeControlPlaneVersion(s)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to compute version of control plane")
+	}
+	if err := contract.ControlPlane().Version().Set(controlPlane, version); err != nil {
 		return nil, errors.Wrap(err, "failed to set spec.version in the ControlPlane object")
 	}
 
 	return controlPlane, nil
 }
 
+// computeControlPlaneVersion returns the desired control plane version.
+// If creating a new control plane, returns the version defined in tolopogy.
+// If a control plane already exists:
+// - If the control plane or any of the machine deployments are upgrading, returns the version from the current control plane.
+// - If not, returns the version defined in topology.
+func computeControlPlaneVersion(s *scope.Scope) (string, error) {
+	topologyVersion := s.Blueprint.Topology.Version
+	// creating a new control plane object, use version from topology
+	if s.Current.ControlPlane == nil || s.Current.ControlPlane.Object == nil {
+		return topologyVersion, nil
+	}
+
+	version, err := contract.ControlPlane().Version().Get(s.Current.ControlPlane.Object)
```

Comment:
nit: let's add a comment here, e.g.

```suggestion
	// Fetch the current version of the control plane object
	version, err := contract.ControlPlane().Version().Get(s.Current.ControlPlane.Object)
```

Reply:
```suggestion
	currentVersion, err := contract.ControlPlane().Version().Get(s.Current.ControlPlane.Object)
```
Also, what about returning earlier if topologyVersion == version?